server/src/main/java/org/elasticsearch/cluster/coordination/RemoveCustomsCommand.java

```diff
@@ -0,0 +1,102 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.elasticsearch.cluster.coordination;
+
+import com.carrotsearch.hppc.cursors.ObjectCursor;
+import joptsimple.OptionSet;
+import joptsimple.OptionSpec;
+import org.elasticsearch.cli.ExitCodes;
+import org.elasticsearch.cli.Terminal;
+import org.elasticsearch.cli.UserException;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.metadata.MetaData;
+import org.elasticsearch.common.collect.Tuple;
+import org.elasticsearch.common.regex.Regex;
+import org.elasticsearch.env.Environment;
+import org.elasticsearch.gateway.PersistedClusterStateService;
+
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.List;
+
+public class RemoveCustomsCommand extends ElasticsearchNodeCommand {
+
+    static final String CUSTOMS_REMOVED_MSG = "Customs were successfully removed from the cluster state";
+    static final String CONFIRMATION_MSG =
+        DELIMITER +
+            "\n" +
+            "You should only run this tool if you have broken custom metadata in the\n" +
+            "cluster state that prevents the cluster state from being loaded.\n" +
+            "This tool can cause data loss and its use should be your last resort.\n" +
+            "\n" +
+            "Do you want to proceed?\n";
+
+    private final OptionSpec<String> arguments;
+
+    public RemoveCustomsCommand() {
+        super("Removes custom metadata from the cluster state");
+        arguments = parser.nonOptions("custom metadata names");
+    }
+
+    @Override
+    protected void processNodePaths(Terminal terminal, Path[] dataPaths, OptionSet options, Environment env)
+        throws IOException, UserException {
+        final List<String> customsToRemove = arguments.values(options);
+        if (customsToRemove.isEmpty()) {
+            throw new UserException(ExitCodes.USAGE, "Must supply at least one custom metadata name to remove");
+        }
+
+        final PersistedClusterStateService persistedClusterStateService = createPersistedClusterStateService(dataPaths);
+
+        terminal.println(Terminal.Verbosity.VERBOSE, "Loading cluster state");
+        final Tuple<Long, ClusterState> termAndClusterState = loadTermAndClusterState(persistedClusterStateService, env);
+        final ClusterState oldClusterState = termAndClusterState.v2();
+        terminal.println(Terminal.Verbosity.VERBOSE, "custom metadata names: " + oldClusterState.metaData().customs().keys());
+        final MetaData.Builder metaDataBuilder = MetaData.builder(oldClusterState.metaData());
+        for (String customToRemove : customsToRemove) {
+            boolean matched = false;
+            for (ObjectCursor<String> customKeyCur : oldClusterState.metaData().customs().keys()) {
+                final String customKey = customKeyCur.value;
+                if (Regex.simpleMatch(customToRemove, customKey)) {
```

Comment:
Should we document that this uses wildcards, e.g. that `*` will remove all customs?

Reply:
This is documented, with the docs saying
> The tool takes a list of custom metadata names as parameters that should be removed, and also supports wildcard patterns.
I don't expect this to be used much, it's more to drop everything as last resort.